### DISCUSSION TITLE
intermittent Messages loss while consuming the messages from consumer

### DISCUSSION THREAD

**Discussion by ragaur-tibco** [apache/pulsar#22681] [Q&A]
Hi @lhotari @visortelle

I have a query
when I tried to run the producer and send 10K messages and then stop the producer and after some time I tried to consume the messages with (Dead letter topic feature but topic is not provided only retry parameters were provided) from the consumer some messages was not present and some of them are duplicate can you please provide your inputs?

```
ConsumerBuilder<String> consumerStringBuilder = client.newConsumer(Schema.STRING)
									        .topic("topic name")
									        .subscriptionName(subscriptionName)
									        .replicateSubscriptionState(false)
									        .subscriptionType(SubscriptionType.Shared);

consumerStringBuilder.negativeAckRedeliveryDelay(60, TimeUnit.SECONDS);
										consumerStringBuilder.deadLetterPolicy(DeadLetterPolicy.builder().maxRedeliverCount(10).deadLetterTopic(null).build());

Consumer<String> stringConsumer= consumerStringBuilder.subscribe();
Message msg = stringConsumer.receive();
```

**Comment by wallacepeng** [comment]
messages should only be retained if there are subscriptions created . from your statement, u first publish messages, then start the consumer, i wonder if lot of messages already deleted ?

**Reply by ragaur-tibco** [reply]
@wallacepeng
I tried with two scenarios 
1. Running the producer and consumer and sending 1 message so that consumer able to consume that 1 message and then stop both producer and consumer then again start the producer and send 10k messages and stop the producer and run the consumer then saw that there are some message loss in while consuming from the consumer 
2. Running the consumer and producer and didn't send any message then stop the consumer and send 10k messages from the producer and then stop the producer and then start the consumer then saw there are some messages loss 

CC: @lhotari @visortelle 

can you please provide any cause of this? and also this issue is intermittent

**Reply by ragaur-tibco** [reply]
@lhotari @wallacepeng 

I just saw one defect on the message loss link: https://github.com/apache/pulsar/issues/22657

in my case I have used the above mentioned points(in my previous comment) after running the consumer to consume the messages saw that there are some loss of messages happening 

Like messages sent 1 to 10k
While running consumer duplication happen for ID's: 557, 880, 960, 990 
Message loss Id: 556 

NOTE: producer is closed after sending 10k messages.  and in standalone.conf file brokerDeduplicationEnabled: "false"

**Reply by ragaur-tibco** [reply]
@lhotari @wallacepeng Also when I restart the consumer then I can able to get the 556 ID in 2nd run

**Reply by lhotari** [reply]
@ragaur-tibco It would be easier to validate your observations if there would be a repro app. You could fork https://github.com/lhotari/pulsar-playground and add a test app in the same way as there are some apps that I have created in the past to investigate some issues. example https://github.com/lhotari/pulsar-playground/blob/master/src/main/java/com/github/lhotari/pulsar/playground/TestScenarioIssue5284.java or more in https://github.com/lhotari/pulsar-playground/tree/master/src/main/java/com/github/lhotari/pulsar/playground . 
You probably already are doing something similar to validate things.

**Reply by lhotari** [reply]
@ragaur-tibco You might find https://github.com/lhotari/pulsar-playground/blob/master/src/main/java/com/github/lhotari/pulsar/playground/TestScenarioIssueRedeliveries.java as a useful example. related to https://github.com/apache/pulsar/pull/21767
It's actually not showing the correct usage of redeliverUnacknowledgedMessages. I'm intentionally testing there to see how things behave if redeliverUnacknowledgedMessages is called very often. That's something that isn't advised. You can take a look at https://github.com/apache/pulsar-site/blob/d7f55bfdb2d6f5f19e5774b4a2d36322f13c9a6e/docs/tutorials-redeliver-messages.md to find some best practices.

**Reply by ragaur-tibco** [reply]
@lhotari, I saw https://github.com/apache/pulsar/pull/21767 , in this you send 10K messages and you mentioned "about 5 to 15 out of 10000 are usually lost." so means there are some messages loss.

**Reply by lhotari** [reply]
> @lhotari, I saw https://github.com/apache/pulsar/pull/21767 , in this you send 10K messages and you mentioned "about 5 to 15 out of 10000 are usually lost." so means there are some messages loss.
> 
> 
> 

I think it was a problem in the test case. I'm not sure if the comments contain that update.

**Reply by lhotari** [reply]
@ragaur-tibco there's [an update comment](https://github.com/apache/pulsar/pull/21767#issuecomment-1867985950) about the mistake in the original test case:

> UPDATE 2: The test case was bad. I had forgotten DLQ config with max redeliveries 5 and that caused the problem.

**Reply by ragaur-tibco** [reply]
@lhotari 

I have used the below config for consumer
"topic"="topic name"
"subscriptionName"=subscriptionName
"replicateSubscriptionState"=false
"subscriptionType"= SubscriptionType.Shared

"negativeAckRedeliveryDelay" = 60
"maxRedeliverCount"=10
"deadLetterTopic"= null

NOTE:  I didn't provide the DLQ but used the below code to receive the messages:

```
>consumerStringBuilder.negativeAckRedeliveryDelay(60, TimeUnit.SECONDS);

> consumerStringBuilder.deadLetterPolicy(DeadLetterPolicy.builder().maxRedeliverCount(10).deadLetterTopic(null).build());

> Consumer<String> stringConsumer= consumerStringBuilder.subscribe();

> Message msg = stringConsumer.receive(); 
```

@lhotari can you please confirm if above use-case is correct or not? 
and can you please confirm do we need to provide the DLQ-topic for above similar use-cases?

NOTE: When the producer and consumer are running simultaneously, the consumer is able to consume all the messages, whereas if the consumer and producer are running separately, the consumer is not able to consume messages there were some messages loss

**Reply by lhotari** [reply]
@ragaur-tibco please share a runnable example that reproduces a problem. I already provided a good template for you so please do your part.

**Reply by lhotari** [reply]
You can take a look at https://github.com/apache/pulsar-site/blob/d7f55bfdb2d6f5f19e5774b4a2d36322f13c9a6e/docs/tutorials-redeliver-messages.md to find some best practices.

**Reply by ragaur-tibco** [reply]
Hi @lhotari 

we have two class one is producer.java and another consumer.java

PulsarProducerExample .java 
=============================================================


```
`import org.apache.pulsar.client.api.*;

public class PulsarProducerExample {
    public static void main(String[] args) throws PulsarClientException {
        String serviceUrl = "pulsar://localhost:6650";
        String topicName = "topic-Name"";

        // Create a Pulsar client instance
        PulsarClient client = PulsarClient.builder()
                .serviceUrl(serviceUrl)
                .build();

        // Create a producer
        Producer<byte[]> producer = client.newProducer()
                .topic(topicName)
                .create();

        // Send messages
        for (int i = 0; i < 100; i++) {
            String message = "Message " + i;
            producer.send(message.getBytes());
            System.out.println("Sent message: " + message);
        }

        // Close the producer and Pulsar client
        producer.close();
        client.close();
    }
}
`
```

Consumer.java
=============================================================

```
`import org.apache.pulsar.client.api.Consumer;
import org.apache.pulsar.client.api.PulsarClient;
import org.apache.pulsar.client.api.SubscriptionType;

public class Consumer{
    public static void main(String[] args) throws Exception {
        String serviceUrl = "pulsar://localhost:6650";
        String topicName = "topic-Name";
        String subscriptionName = "subscriptionName";

        PulsarClient client = PulsarClient.builder()
                .serviceUrl(serviceUrl)
                .build();

        Consumer<byte[]> consumer1 = client.newConsumer()
                .topic(topicName)
                .subscriptionName(subscriptionName)
                .subscriptionType(SubscriptionType.Shared) // Use Shared subscription type
                .subscribe();
        while (true) {
           Message msg = consumer1.receive();
            System.out.println("from consumer 1=============================="+consumer1.receive());
            consumer1.acknowledgeAsync(msg);
        }
    }
}



`
```

Steps to reproduce:

1. Run consumer and topic will be created by itself
2. check the topic stats one consumer created in subscription name
3. stop the consumer and then run the producer and send 100 messages 
4. Run consumer and check the topic stats now you will be able to see that a different consumer with different name and some time with same name created (if in consumerName parameter provided with the same consumer name which is already present) and 100 messages available in previous consumer but not in new consumer. Due to which not able to receive messages

Queries:
1. Is there any api from which we can receive the messages from the specific consumer name
2. Is there any way to fetch the all the messages with is available in subscription from all the consumers

**Reply by lhotari** [reply]
> 4\. Run consumer and check the topic stats now you will be able to see that a different consumer with different name and some time with same name created (if in consumerName parameter provided with the same consumer name which is already present) and 100 messages available in previous consumer but not in new consumer. Due to which not able to receive messages

how is this step reflected in the above code for reproducing the issue?
I don't see any issue here. The code has a shared subscription and the subscription name is the same. Each consumer must have a unique name if you'd like to have the messages delivered to this "different consumer with different name". 

> Is there any api from which we can receive the messages from the specific consumer name

No. The consumer name isn't used for receiving messages. 

> Is there any way to fetch the all the messages with is available in subscription from all the consumers

I don't understand this question. What are you trying accomplish here? For each subscription there's a matching cursor which has runtime state that can be observed with the topic [statistics](https://pulsar.apache.org/docs/3.2.x/admin-api-topics/#get-stats) ([reference](https://pulsar.apache.org/docs/3.2.x/administration-stats/#topic-stats)) and [internal statistics](https://pulsar.apache.org/docs/3.2.x/admin-api-topics/#get-internal-stats) ([reference](https://pulsar.apache.org/docs/3.2.x/administration-stats/#topic-internal-stats)). Some of the cursor internal state is the so called "mark delete position", "read position" and "individually deleted messages". Deletion means acknowledgement in these concepts. The "mark delete position" is the point where all previous messages have been acknowledged in the cursor (and subscription). For more details you could check this blog post: https://dzone.com/articles/understanding-cursors-tracking-mechanism-in-pulsar .
